ykcompile/src/lib.rs

```diff
@@ -383,49 +460,45 @@ impl<TT> TraceCompiler<TT> {
                     _ => todo!("{}", ty.size()), // FIXME: For things >16, use memcpy().
                 }
             }
-            (Location::Register(reg), Location::Deref(boxed)) => match ty.size() {
-                8 => match *boxed {
-                    Location::Mem(ro) => {
-                        dynasm!(self.asm
-                            ; mov Rq(reg), [Rq(ro.reg) + ro.offs]
-                            ; mov Rq(reg), [Rq(reg)]
-                        );
-                    }
-                    Location::Register(reg2) => {
-                        dynasm!(self.asm
-                            ; mov Rq(reg), [Rq(reg2)]
-                        );
-                    }
-                    _ => todo!(),
-                },
-                _ => todo!("{}", ty.size()),
-            },
-            (Location::Mem(ro1), Location::Deref(boxed)) => match ty.size() {
-                8 => match *boxed {
-                    Location::Mem(ro2) => {
-                        dynasm!(self.asm
-                            ; mov rax, [Rq(ro2.reg) + ro2.offs]
-                            ; mov rax, [rax]
-                            ; mov [Rq(ro1.reg) + ro1.offs], rax
-                        );
-                    }
-                    Location::Register(reg) => {
-                        dynasm!(self.asm
-                            ; mov rax, [Rq(reg)]
-                            ; mov [Rq(ro1.reg) + ro1.offs], rax
-                        );
-                    }
-                    _ => todo!(),
-                },
-                _ => todo!("{}", ty.size()),
-            },
             _ => unreachable!(),
         }
+
+        // Free temporary if one was created.
+        self.free_temporary(rloc);
         Ok(())
     }
 
     fn mov_place_ref(&mut self, p1: &Place, p2: &Place) -> Result<(), CompileError> {
         let lloc = self.place_to_location(p1)?;
+
+        // Deal with the special case `&*`, i.e. referencing a `Deref` on a reference just returns
+        // the reference.
```

Comment:
Is that still true in light of custom `Deref` impls? Do we need a fixme?

Reply:
Custom `Deref` impls are desugared when creating MIR.

Reply:
Do they become calls?

Reply:
Yes

Reply:
```rust
fn main() {
    &*Vec::<()>::new();
}
```

```rust
// WARNING: This output format is intended for human consumers only
// and is subject to change without notice. Knock yourself out.
fn main() -> () {
    let mut _0: ();
    let _1: &[()];
    let mut _2: &std::vec::Vec<()>;
    let _3: std::vec::Vec<()>;

    bb0: {
        StorageLive(_1);
        StorageLive(_2);
        StorageLive(_3);
        _3 = const std::vec::Vec::<()>::new() -> bb2;
    }

    bb1 (cleanup): {
        resume;
    }

    bb2: {
        _2 = &_3;
        _1 = const <std::vec::Vec<()> as std::ops::Deref>::deref(move _2) -> [return: bb3, unwind: bb4];
    }

    bb3: {
        StorageDead(_2);
        drop(_3) -> bb5;
    }

    bb4 (cleanup): {
        drop(_3) -> bb1;
    }

    bb5: {
        StorageDead(_3);
        StorageDead(_1);
        _0 = const ();
        return;
    }
}
```

Reply:
> Is that still true in light of custom `Deref` impls? Do we need a fixme?

Note that I check that the place `p2` is a reference. Dereferencing a reference apparently does not trigger custom `Deref` operations.

> Custom Deref impls are desugared when creating MIR.

That's a very useful information that hopefully makes our lives easier.